src/main/java/org/testng/internal/TestResult.java

```diff
@@ -87,7 +87,7 @@ public void init (IClass testClass,
       if (m_instance instanceof ITest) {
         m_name = ((ITest) m_instance).getTestName();
       }
-      else if (testClass.getTestName() != null && !testClass.getTestName().isEmpty()) {

```

Comment:
Added. Funny enough, during testing I discovered that I was wrong about the issue with blank test names, because the check for ITest comes before this. D'oh. Doesn't hurt to have more tests, though!